### ISSUE TITLE
Gradient Clipping

### ISSUE THREAD

**Issue by AlbertoSabater** [davda54/sam#32]
What is the proper way to perform gradient clipping with the SAM/ASAM optimizer? Should it be performed before both of the optimizer steps or just one of them?

Code example:

```
  # first forward-backward pass
  loss = loss_function(output, model(input))  # use this loss for any training statistics
  loss.backward()
  torch.nn.utils.clip_grad_norm_(model.parameters(), v)
  optimizer.first_step(zero_grad=True)
  
  # second forward-backward pass
  loss_function(output, model(input)).backward()  # make sure to do a full forward pass
  torch.nn.utils.clip_grad_norm_(model.parameters(), v)
  optimizer.second_step(zero_grad=True)
```

**Comment by JOBR0** [comment]
The clipping before the first step should not have any effect because the gradient is divided by its norm in the first step anyway.
